### PR TITLE
Add Shared Library Tool

### PR DESCRIPTION
**Pull request by gmittert**
Added a tool to link and produce .dll/.so outputs.

I'm planning on using this in SwiftPM for linking Windows Swift code since Windows Swift currently doesn't support static linking. It currently works for that purpose and produces .dlls and import libs on Windows which I've used to (partially) compile SwiftPM.

### CODE REVIEW

File: lib/BuildSystem/BuildSystem.cpp [hunk 1]
```diff
@@ -3320,6 +3320,195 @@ class ArchiveTool : public Tool {
...
+        args.push_back(clangExecutable);
+      } else if (language == "C++") {
+        args.push_back(clangPlusPlusExecutable);
+      }
+      args.insert(args.end(), sharedLibInputs.begin(), sharedLibInputs.end());
+      args.push_back("-o");
+      args.push_back(sharedLibName);
+      args.insert(args.end(), otherArgs.begin(), otherArgs.end());
+
+#if defined(_WIN32)
```
Comment (gmittert) [inline lib/BuildSystem/BuildSystem.cpp]: Do we care about cross compilation? If so, I'll add a target parameter and base the options added on that.

Comment (jakepetroules) [reply lib/BuildSystem/BuildSystem.cpp]: Yes, please do so. LLVM philosophy :)

File: docs/buildsystem.rst [hunk 2]
```diff
@@ -785,6 +785,36 @@ A typical use for this tool is creating static libraries.
...
+
+   * - swiftExecutable
+     - A string indicating the path to a ``swiftc`` compiler that will be used
+       to link Swift code.
+
+   * - clangExecutable
+     - A string indicating the path to a ``clang`` compiler that will be used
+       to link C code.
+
+   * - clang++Executable
```
Comment (jakepetroules) [inline docs/buildsystem.rst]: It feels weird having "+" in an identifier, can you please use something like "clangPlusPlusExecutable" instead?

File: lib/BuildSystem/BuildSystem.cpp [hunk 3]
```diff
@@ -3320,6 +3320,195 @@ class ArchiveTool : public Tool {
   }
 };
 
+#pragma mark - SharedLibraryTool implementation
+
+class SharedLibraryShellCommand : public ExternalCommand {
+
+  std::string swiftExecutable = "swiftc";
+#if defined(_WIN32)
```
Comment (jakepetroules) [inline lib/BuildSystem/BuildSystem.cpp]: Please add a comment that these are only defaults (to make clear that the compile time check for host OS isn't preventing cross compilation).

**Comment by gmittert** [comment]
Huh, so turns out detecting what compiler you've been given is... somewhat difficult... to say the least. Rather than trying to guess the compiler by its executable name or determining what it is by running it. I've just added a "compiler-style" field when can be specified to be "cl, clang, or swiftc".

**Comment by gmittert** [comment]
<OneWeekCourtesyPing>

**Comment by gmittert** [comment]
@swift-ci please test

**Comment by gmittert** [comment]
@swift-ci please smoke test

**Comment by gmittert** [comment]
rebased and addressed feedback

@swift-ci please smoke test

**Comment by gmittert** [comment]
@jakepetroules Re I believe I've addressed your requested changes.

**Pull request merged**